tasks/upsert_component.py

```diff
@@ -0,0 +1,138 @@
+    value: float
+
+
+def create_measurement_dict(
+    name: str, commit: Commit, measurable_id: str, value: float
+) -> MeasurementDict:
+    return {
+        "name": name,
+        "owner_id": commit.repository.ownerid,
+        "repo_id": commit.repoid,
+        "measurable_id": measurable_id,
+        "branch": commit.branch,
+        "commit_sha": commit.commitid,
+        "timestamp": commit.timestamp,
+        "value": value,
+    }
+
+
+def upsert_measurements(
+    db_session: Session, measurements: list[MeasurementDict]
+) -> None:
+    command = insert(Measurement.__table__).values(measurements)
+    command = command.on_conflict_do_update(
+        index_elements=[
+            Measurement.name,
+            Measurement.owner_id,
+            Measurement.repo_id,
+            Measurement.measurable_id,
+            Measurement.commit_sha,
+            Measurement.timestamp,
+        ],
+        set_=dict(
+            branch=command.excluded.branch,
+            value=command.excluded.value,
+        ),
+    )
+    db_session.execute(command)
+    db_session.flush()
+
+
+class UpsertComponentTask(BaseCodecovTask):
+    def run_impl(
+        self,
+        db_session: Session,
+        commitid: str,
+        repoid: int,
+        component_id: str,
+        *args,
+        **kwargs,
+    ):
+        log.info("Upserting component", extra=dict(commitid=commitid, repoid=repoid))
+
+        commit = (
+            db_session.query(Commit)
+            .filter(Commit.repoid == repoid, Commit.commitid == commitid)
+            .first()
+        )
+
+        installation_name_to_use = get_installation_name_for_owner_for_task(
+            self.name, commit.repository.owner
+        )
+
+        repository_service = get_repo_provider_service(
+            commit.repository, installation_name_to_use=installation_name_to_use
+        )
+
+        yaml: UserYaml = async_to_sync(get_current_yaml)(commit, repository_service)
+
+        components = yaml.get_components()
+
+        report_service = ReportService(yaml)
+        report = report_service.get_existing_report_for_commit(
+            commit, report_class=ReadOnlyReport
+        )
+
+        if report is None:
+            log.warning(
+                "Upsert Component: No report found for commit",
+                extra=dict(
+                    component_id=component_id,
+                    commitid=commitid,
+                    repoid=repoid,
+                ),
+            )
+            return
+
+        component_dict = {component.component_id: component for component in components}
+
+        component = component_dict[component_id]
+
+        if component.paths or component.flag_regexes:
+            report_and_component_matching_flags = component.get_matching_flags(
+                list(report.flags.keys())
+            )
+            filtered_report = report.filter(
+                flags=report_and_component_matching_flags, paths=component.paths
+            )
```

Comment:
I believe the primary slowness comes from the `report.filter` call, which is the thing we would like to parallelize.

All the rest should be trivial and fast. And I would imagine that the overhead of task spawning, fetching the report, and fetching the yaml is quite a bit of overhead.

How about:
You do the deduplication, resolution and filtering of these components in the main task.
As the main task is fetching the report anyway, you can end up with a list of `component_id, flags, paths` which are already filtered down to only the relevant ones.

Then when you have that filtered and resolved list, you either iterate over that list, or spawn individual tasks.